regression-tests/pure2-last-use.cpp2

```diff
@@ -0,0 +1,1044 @@
+    _ = e is (e);
+
+    f := new<int>(0);
+    _ = f is std::type_identity_t<decltype(f)>; // OK?
+
+    g := new<int>(0);
+    for (identity(g)* + identity(g)*)
+    do (_)
+    { }
+
+    h := new<int>(0);
+    while identity(h)* + identity(h)*
+    { }
+
+    i := new<int>(0);
+    do { }
+    while identity(i)* + identity(i)*;
+
+    j := new<int>(0);
+    k := new<int>(0);
+    _ = inspect identity(j)* + identity(j)* -> int {
+        is (identity(k)* + identity(k)*) = 0;
+        is _ = 0;
+    };
+
+  }
+  {
+    a := new<int>(0);
+    _ = :() identity(a$)* + identity(a$)*;
+  }
+}
+issue_313_1: (copy x: std::unique_ptr<int>) pre(identity(x)* + identity(x)*) = { }
+
+issue_350: () = {
+    x := 21;
+
+    l1 := :(forward x) = {
+        f_inout(forward x);
+    };
+
+    l1(x);
+
+    x++;
+}
+/*
+issue_440_0: () -> int = {
+  i: int;
+  if true {
+    i = 1;
+    return i;
+  }
+  i = 2;
+  return i;
+}
+
+issue_440_1: () -> (i: int) = {
+  if true {
+    i = 1;
+    return;
+  }
+  i = 2;
+}
+*/
+issue_683: (args) = {
+    for args do (n) {
+        _ = n;
+    }
+
+    n: int;
+    n = 0;
+}
+
+issue_825: () = {
+  _ = :(copy b: std::unique_ptr<int>) f_copy(b);
+  _ = :(move c: std::unique_ptr<int>) f_copy(c);
+  _ = :(forward d) f_copy(d);(new<int>(0));
+}
+
+issue_832: () = {
+  i := 0;
+  while i { }
+}
+
+make_copy: (copy x) x;
+
+issue_847_4: (copy v: std::vector<int>) = {
+  for v.make_copy()
+  do (copy x) {
+    f_copy(x);
+  }
+}
+issue_847_0: (copy v: std::vector<std::unique_ptr<int>>) = {
+  // TODO Use 'std::views::as_rvalue`
+  for v.make_copy()
+  do (_) {
+  }
+}
+issue_847_1: (move v: std::vector<std::unique_ptr<int>>) = {
+  for v.make_copy()
+  do (move x) {
+    f_copy(x);
+  }
+}
+issue_847_5: (forward v) = {
+  for v.make_copy()
+  do (forward x) {
+    f_copy(x);
+  }
+}
+issue_847_2: (forward v) = {
+  for v.make_copy()
+  do (forward _) {
+  }
+}
+issue_847_3: (copy x: int) = { for (x) do (_) { } }
+
+issue_850: () = {
+    v: std::vector = ( 1, 2, 3, 4, 5 );
+
+    //  Definite last use of v => move-capture v into f's closure
+    f := :() -> forward _ = { return v$; };
+
+    //  Now we can access the vector captured inside f()...
+    f().push_back(6);
+    for f() do(e) std::cout << e;       // prints 123456
+}
+
+issue_857: type = {
+  a: std::unique_ptr<int>;
+  b: std::unique_ptr<int>;
+  operator=: (out this, move that) = { }
+  operator=: (move this) = {
+    f_inout(a);
+    f_copy(this.b);
+  }
+  f: (move this) = f_copy(this);
+  f: (move this, move that) = f_copy(this, that);
+  g: (move this) = f_copy(this.a);
+  g: (move this, move that) = f_copy(this.a, that.a);
+  h: (inout this) = f_inout(a);
+  i: (move this) = f_copy(a);
+  j: (move this) = f_copy(a);
+  k: (move this) = {
+    f_inout(a);
+    f_copy(b);
+  }
+  l: (move this) = k();
+  m: (move this) = this.k();
+  n: (_) = { }
+  n: (move this) = { }
+  o0:(move this) = n();
+  o1:(move this) = this.n();
+  o2:(move this) = 0.n();
```

Comment:
For this line, when I tried compiling the test using cppfront+MSVC, cppfront was fine but MSVC complained:

```
pure2-last-use.cpp2(193): error C3889: call to object of class type 'issue_857::o2::<lambda_1>': no matching call operator found
pure2-last-use.cpp2(193): note: could be 'decltype(auto) issue_857::o2::<lambda_1>::operator ()(Obj &&,Params &&...) noexcept(<expr>) const'
pure2-last-use.cpp2(193): note: the associated constraints are not satisfied
pure2-last-use.cpp2(193): note: the constraint was not satisfied
```

Reply:
Dropping the UFCS macro from `CPP2_UFCS(std::move(*this).n)(0);`
makes GCC and MSVC work: <https://cpp1.godbolt.org/z/fWsWxbM34>.
When lookup finds the member `n`, it won't continue past the type scope.
So we can just disable UFCS in this case: <https://cpp2.godbolt.org/z/dax8Psaqj>.

Reply:
Ah, but then this would attempt to use the non-existent Cpp1 UFCS on a literal: `0.n()`.